### PR TITLE
Increase speed of augmented lookup for `AugmentedMemoizationPolicy`

### PR DESCRIPTION
**Pull request by JEM-Mosig**
**Proposed changes**:
- Fixes https://github.com/RasaHQ/rasa-x/issues/5546
- Copies essential changes from https://github.com/RasaHQ/rasa/pull/6711 to this old Rasa version

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] ~updated the documentation~
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by JEM-Mosig** [comment]
@wochinge I ported my fix to 1.10.x compatibility. I had to add `**kwargs` to the trackers so `TrackerWithCachedStates` can be given the domain with `from_events` in the test code. I also copied over the much of the comments for `AugmentedMemoizationPolicy` that are present in `2.x` but not in `1.x`.

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py]: isn't max history related to `ActionExecuted` and not all tracker events?

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py:324]: optional: why do we need to pass over `max_history`? `self._back_to_the_future` should have access to it as well through `self.max_history`, no?

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py:296]: ? 🤔 
```suggestion
        events = applied_events[start_index + idx_to_use:]
```

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py:276]: why don't we do this once in `recall` instead of here where this is part of the loop?

File: rasa/core/trackers.py
Comment (wochinge) [inline rasa/core/trackers.py:115]:
```suggestion
            **kwargs: Any,
```

File: rasa/core/training/generator.py
Comment (wochinge) [inline rasa/core/training/generator.py:47]: how about overriding `from_events` in here instead doing changes to the `DialogueStateTracker` which doesn't need the domain?

**Review by wochinge — COMMENTED** [review]
Thanks for the PR and your work! 💯 
Did we also do a benchmark to whether it actually fixes things?

**Comment by wochinge** [comment]
And another researcher as reviewer (maybe just for the final one) would probably be a good idea 🙌🏻

File: rasa/core/policies/memoization.py
Comment (JEM-Mosig) [reply rasa/core/policies/memoization.py]: Damn. Don't know what I was thinking. Maybe the vaccine is affecting my mind after all :(

File: rasa/core/policies/memoization.py
Comment (JEM-Mosig) [reply rasa/core/policies/memoization.py:324]: `_back_to_the_future_again` is static, but I can change that.

File: rasa/core/policies/memoization.py
Comment (JEM-Mosig) [reply rasa/core/policies/memoization.py:296]: 🤦

**Comment by JEM-Mosig** [comment]
> Did we also do a benchmark to whether it actually fixes things?
Yes. Introducing the change reduces the prediction time from 4.68 s to 0.003 s on a 1000 event tracker.

**Comment by wochinge** [comment]
Just re-request review once it's ready for another review 👍🏻

**Comment by dakshvar22** [comment]
Don't think I have the capacity to review but just a reminder that we should run model regression tests for core before merging any of these PRs (I guess there are multiple of them for different versions?)

File: tests/core/test_processor.py
Comment (kedz) [inline tests/core/test_processor.py:648]: What's the reason for keeping these comments in here?

Comment (JEM-Mosig) [reply tests/core/test_processor.py:648]: The comments show how this changes with 2.x

File: tests/core/test_processor.py
Comment (kedz) [inline tests/core/test_processor.py:672]: What's the reason for keeping these comments in here? Not sure what you are trying to communicate with these.

**Review by kedz — COMMENTED** [review]
Implementation and tests looks good. I had a clarification question about some of the comments in the tests though that should get answered before this is approved.

File: tests/core/test_processor.py
Comment (kedz) [reply tests/core/test_processor.py:648]: Could that be stated somewhere or is this a practice that is in use throughout the code base and I am just unaware of it?

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py:375]: Optional: Let's make this a protected method
```suggestion
def _get_max_applied_events_for_max_history(
```

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py:328]: why do we need the extra param? We could also create a new (shorter) tracker right away and keep the rest of the code the same?

File: tests/core/test_processor.py
Comment (JEM-Mosig) [reply tests/core/test_processor.py:648]: You're right, I'll add this to the comment line.

Comment (wochinge) [reply tests/core/test_processor.py:648]: I'd vote to remove this 👍🏻

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py:291]: nit: `return None` is actually preferable as this is more explicit (especially as you already have other return statements which return values)

File: rasa/core/policies/memoization.py
Comment (wochinge) [inline rasa/core/policies/memoization.py:302]: what does this sentence mean? 🙈

File: rasa/core/training/generator.py
Comment (wochinge) [inline rasa/core/training/generator.py:66]: return type is missing

File: rasa/core/policies/memoization.py
Comment (JEM-Mosig) [reply rasa/core/policies/memoization.py:302]: It is directly copied from main. I'd rewrite much of the names and comments, but this is an old version

**Comment by wochinge** [comment]
@JEM-Mosig I'll be on vacation next week so please request review from somebody else then in case you need another review from an engineer

**Review by joejuzl — APPROVED** [review]
LGTM 👍

**Review by kedz — APPROVED** [review]
LGTM2! 🚀

**Pull request merged**
